[[ LC Builder ]] Implement various things in encoding module

livecodeali opened the pull request.

Commit:
[[ LC Builder ]] Implement various things in encoding module

Conflicts:
	engine/src/modules.cpp
	libscript/libscript.xcodeproj/project.pbxproj
	libscript/libstdscript-modules.list
	toolchain/lc-compile/lc-compile.xcodeproj/project.pbxproj
	toolchain/lc-compile/src/module-helper.cpp

Commit:
[[ LC Builder ]] Fix typo in encode using base64 docs

peter-b commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -501,6 +501,98 @@ bool MCStringCreateMutable(uindex_t p_initial_capacity, MCStringRef& r_string)
+
+// These names must match the formatting of character encoding names as output
+// by the algorithm described in section 1.4 of the Unicode Consortium TR22.
+static MCStringEncodingName MCStringEncodingNames[] =
```
Do you think it would be a good idea to add the "official" hyphenated versions of the encoding names as synonyms, e.g.:

```
iso-8859-1
utf-8
```

etc.?

runrevmark replied:
I think this covered by the standard name processing that occurs - Fraser can confirm :)

Sent from my iPhone

> On 16 Jan 2015, at 08:56, Peter TB Brett notifications@github.com wrote:
> 
> In libfoundation/src/foundation-string.cpp:
> 
> > @@ -501,6 +501,98 @@ bool MCStringCreateMutable(uindex_t p_initial_capacity, MCStringRef& r_string)
> > 
> >  ////////////////////////////////////////////////////////////////////////////////
> > 
> > +struct MCStringEncodingName
> > +{
> > -    const char *m_name;
> > -    MCStringEncoding m_encoding;
> >   +};
> >   +
> >   +// These names must match the formatting of character encoding names as output
> >   +// by the algorithm described in section 1.4 of the Unicode Consortium TR22.
> >   +static MCStringEncodingName MCStringEncodingNames[] =
> >   Do you think it would be a good idea to add the "official" hyphenated versions of the encoding names as synonyms, e.g.:
> 
> iso-8859-1
> utf-8
> etc.?
> 
> —
> Reply to this email directly or view it on GitHub.

livecodeali commented:
This currently crashes when concatenating a list element with a string literal and passing it through to a handler.

Commit:
[[ LCB StdLib ]] Remove syntactic forms of text encodings

peter-b commented:
Is this a good time to address http://quality.runrev.com/show_bug.cgi?id=13892 by making the builder encoding/decoding syntax return kMCNull if a codec error occurs? I.e.

```
encode "∞" using "iso-8859-1"      -- Result is undefined, not "?"
```

This would be consistent with the behaviour agreed on for `parse`... :wink:

Commit:
[[ LCB StdLib ]] Update lc-compile project file

Commit:
[[ LCB StdLib ]] Allow replacement byte (resp. char) to be specified when encoding (resp. decoding)
[[ LCB StdLib ]] Return undefined when replacement is unspecified and encoding or decoding fails

livecodeali commented:
We need to decide what happens when a replacement character is specified but the data is invalid due to length, eg decoding 3 bytes from utf32 or 1 from utf16. Currently these are decoded as empty.

runrevmark commented:
I think those count the same as if you had an invalid sequence in utf-8. When decoding, if you specify a replacement char then it is by definition for invalid byte sequences (notionally our internal rep - unicode - is a superset of all encodings so the non-encodable character case doesn't come up).

I'm still on the fence about utf-16 encoded as utf-8 and isolated surrogates in utf-16. Some systems (Java for example) do generate utf-8 with surrogates which is (strictly speaking) incorrect but there's the interoperability aspect to consider (perhaps this case could be considered as a different encoding?)

On the utf-16 front - since we allow code unit access, LC handles isolated surrogates fine so there isn't necessarily a problem. Again, maybe this could be handled via a separate encoding. (The same is true of utf-32)

Sent from my iPhone

> On 23 Jan 2015, at 12:49, Ali Lloyd notifications@github.com wrote:
> 
> We need to decide what happens when a replacement character is specified but the data is invalid due to length, eg decoding 3 bytes from utf32 or 1 from utf16. Currently these are decoded as empty.
> 
> —
> Reply to this email directly or view it on GitHub.

Commit:
[[ LCB StdLib ]] Deal with invalid data appropriately

Commit:
[[ LBC StdLib ]] Add encoding module to vs build rules

Commit:
[[ LCB StdLib ]] Allow replacement char for invalid utf16 and utf32 byte sequences

peter-b commented on libfoundation/include/foundation.h:
```diff
@@ -1748,15 +1756,28 @@ MC_DLLEXPORT bool MCStringCreateMutable(uindex_t initial_capacity, MCStringRef&
 MC_DLLEXPORT bool MCStringEncodeAndRelease(MCStringRef string, MCStringEncoding encoding, bool is_external_rep, MCDataRef& r_data);
+    
+// Encode the given string with the specified encoding. Characters which cannot
+// be represented in the target encoding are replaced by the replacement byte, or,
```
"replacement bytes"

peter-b commented on libfoundation/include/foundation.h:
```diff
@@ -1748,15 +1756,28 @@ MC_DLLEXPORT bool MCStringCreateMutable(uindex_t initial_capacity, MCStringRef&
 MC_DLLEXPORT bool MCStringDecodeAndRelease(MCDataRef data, MCStringEncoding encoding, bool is_external_rep, MCStringRef& r_string);
 
+// Decode the given data, intepreting in the given encoding. Invalid bytes for the
+// target encoding are replaced by the replacement byte, or, if p_replacement is nil,
```
"replacement string"

peter-b commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -1111,28 +1442,46 @@ uindex_t MCStringGetChars(MCStringRef self, MCRange p_range, unichar_t *p_chars)
+        {
+            if (!MCUnicodeCharMapToNative(self -> chars[i], t_native_char))
+            {
+                if (p_replacement)
```
Having `p_replacement` passed in as a pointer suggests that `p_replacement` could be a multi-byte sequence (e.g. `BAD`).

If `*p_replacement` is `\0`, consider abandoning the loop early.

peter-b committed:
Merge branch 'develop' into builder-encoding_module

peter-b committed:
libfoundation: Remove MCStringGetNativeChars() from public API.

Remove both MCStringGetNativeChars() and
MCStringGetNativeCharsWithReplacement() from the public libfoundation
API.

peter-b committed:
libfoundation: Update comments for MCStringConvertToBytesWithReplacement().

peter-b committed:
[[ LCB stdlib ]] Permit multi-byte replacements.

peter-b committed:
libfoundation: Update comments for MCStringEncodeWithReplacement().

peter-b committed:
libfoundation: Refactor MCStringGetNativeChars() API.

* Return true on success/false on failure from
  MCStringGetNativeChars() and
  MCStringGetNativeCharsWithReplacement().

* Add an in parameter for size of character buffer provided and an out
  parameter for number of characters used/needed.

* Allow multi-byte replacement sequences for characters that can't be
  represented in the native encoding, and add an explicit in parameter
  for the replacement sequence length.

peter-b committed:
libfoundation: Refactor MCStringConvertToNativeWithReplacement().

Allow zero-byte and multibyte replacement sequences.  Note that it is
*not* an error for the replacement array to contain non-ASCII values.

peter-b commented:
Still need to look into decoding with multibyte replacements.

peter-b commented:
The story continues elsewhere...
